src/game/connect4.rs

```diff
@@ -42,43 +60,249 @@ impl GameAdapter for Connect4Adapter<'_> {
-            self.players.push(username);
-            if self.players.len() == NUM_PLAYERS {
-                self.state = State::InProgress
-            }
-            self.notifier.send(());
-            Ok(())
-        } else {
-            Err(actix_web::Error::from(
+        if self.players.len() >= NUM_PLAYERS {
+            return Err(actix_web::Error::from(
                 GameAdapterError::PlayerLimitExceeded(NUM_PLAYERS),
-            ))
+            ));
+        }
+        self.players.push(username);
+        if self.players.len() == NUM_PLAYERS {
+            self.state = State::InProgress
         }
+        self.notifier.send(());
+        Ok(())
     }
 
     fn has_player(&self, username: &str) -> bool {
         self.players.iter().any(|s| s.eq(username))
     }
 
+    fn get_user_from_token(&self) -> String {
+        let user = match self.game.turn {
+            Token::Red => self.players.get(0).unwrap().clone(),
+            Token::Blue => self.players.get(1).unwrap().clone(),
+        };
+        user
+    }
+
     fn play_move(&mut self, game_move: GenericGameMove) -> actix_web::Result<()> {
         if self.state != State::InProgress {
             return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
                 self.state,
             )));
-        };
-        todo!()
+        }
+        let request_payload = serde_json::from_value::<Connect4RequestPayload>(game_move.payload)?;
+        let column = request_payload.column;
+        let player = self.get_user_from_token();
+        let user = game_move.player;
+
+        if player != user {
+            return Err(actix_web::Error::from(
+                GameAdapterError::WrongPlayerRequest(user),
+            )); // return the one who made the request
+        }
+        self.game.moves(column)?;
+        let win = self.game.winning_move(column);
+        let draw = self.game.is_game_drawn();
+        if win {
+            self.winner.push(self.get_user_from_token());
+        }
+        if win || draw {
+            self.game.completed = true;
+        } else {
+            self.game.switch_token();
+        }
+        self.notifier.send(());
+        Ok(())
     }
 
     fn get_encoded_state(&self) -> actix_web::Result<GenericGameState> {
+        let encoded_board = self
+            .game
+            .board
+            .iter()
+            .map(|col| {
+                col.iter()
+                    .map(|&token| match token {
+                        Token::Red => self.players.get(0),
+                        Token::Blue => self.players.get(1),
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
-            can_move: vec![self.next_move.clone()],
+            can_move: vec![self.get_user_from_token()],
```

Comment:
Do not call this if the game state isn't `InProgress`, just return an empty vector instead